builder/builder.go

```diff
@@ -199,15 +204,143 @@ func (b *Builder) startRegistryProxy(ctx context.Context, caps apicaps.CapSet) (
 		doneCh <- struct{}{}
 	}()
 
+	isDarwin := runtime.GOOS == "darwin"
+
+	if isDarwin {
+		darwinProxyPort, err := b.startDarwinProxyHelper(ctx, registryPort)
+		if err != nil {
+			b.opt.Console.Warnf("Failed to start registry proxy support container: %v", err)
+			err := b.stopDarwinProxyHelper(ctx) // Clean up dangling container on failure.
+			if err != nil {
+				b.opt.Console.Warnf("Failed to stop registry proxy support container: %v", err)
+			}
+			return nil, false
+		}
+		b.opt.Console.VerbosePrintf("Started support container for registry proxy on port %d", darwinProxyPort)
+		addr = fmt.Sprintf("127.0.0.1:%d", darwinProxyPort)
+	}
+
 	b.opt.LocalRegistryAddr = addr
 
-	return func() {
+	closeFn := func() {
+		fmt.Println("Closing registry proxy")
+		if isDarwin {
+			err := b.stopDarwinProxyHelper(ctx)
+			if err != nil {
+				b.opt.Console.Warnf("Failed to stop registry proxy support container: %v", err)
+			}
+		}
 		p.Close()
 		select {
 		case <-ctx.Done():
 		case <-doneCh:
 		}
-	}, true
+	}
+
+	return closeFn, true
+}
+
+const darwinProxyContainerName = "darwin-registry-proxy"
+
+// startDarwinProxyHelper: Since Docker Desktop (Mac) containers run in a VM, a
+// special host name, host.docker.internal, is made available to access the
+// host's network. Docker can only pull insecurely from localhost, so we use a
+// socat container to proxy localhost:<port> request back out to the local
+// registry proxy created above.
+func (b *Builder) startDarwinProxyHelper(ctx context.Context, registryPort int) (int, error) {
+	infoMap, err := b.opt.ContainerFrontend.ContainerInfo(ctx, darwinProxyContainerName)
```

Comment:
This can present issues when running Earthly in parallel on the same system, right? We might need to come up with a unique name here.

Reply:
Good point. I'll use a dynamic name.